src/execute.c

```diff
@@ -921,6 +1096,243 @@ jv jq_next(jq_state *jq) {
+       *
+       * But first we'll push a value so that when the unwind-protect handler
+       * backtracks we then backtrack too.
+       */
+      if (jv_get_kind(done) == JV_KIND_FALSE) {
+        if (!raising) {
+          /*
+           * We push true because when we backtrack here it will be because an
+           * unwind-protect handler is done and so we'll take the else brack if
+           * the if jv_get_kind(done) == JV_KIND_FALSE) above.
+           */
+          stack_push(jq, jv_true());
+        } else {
+          /*
+           * We'll push the error's message, and we wrap it in an array in case
+           * the message was a jv_true().
+           */
+          stack_push(jq, JV_ARRAY(jv_invalid_get_msg(jq->error)));
+          jq->error = jv_null();
+        }
+        stack_save(jq, pc - 1, stack_get_pos(jq));
+      } else {
+        /*
+         * We're done then.  If done is an array, the one element will be an
+         * error message.
+         */
+        if (jv_get_kind(done) == JV_KIND_ARRAY)
+          set_error(jq, jv_invalid_with_msg(jv_array_get(done, 0)));
+        goto do_backtrack;
+      }
+      break;
+    }
+
+    case COCREATE: {
+      jv input = stack_pop(jq);
+      if (jv_get_kind(input) == JV_KIND_NULL) {
+        /* Here we're in the parent */
+        /*
+         * Fork a child co-routine.
+         *
+         * The child side will see `. == false`.  The parent will see an open I/O
+         * handle corresponding to the child side.  If this seems familiar, it's
+         * because this is a lot like the Unix fork(2) system call.
+         *
+         * coinit() makes a copy of the parent jq_state, but arranges for the
+         * child never to stack_restore() past this point.  See jq_reset().
+         *
+         * This all works because we never store pointers into the stack on the
+         * stack without first encoding them to be relative to the stack, and
+         * because the child cannot backtrack past this point, the parent will
+         * teardown the child if the parent backtracks past this point, and there
+         * is no concurrent access to the stack.
+         */
+        jq_state *child = coinit(jq);
+        /*
+         * We push a fork point on the child stack that the child co-routine can
+         * be restored up to here, start from here, and when it runs out of
+         * values it can stop here.
+         */
+        stack_push(child, jv_true());
+        stack_save(child, pc - 1, stack_get_pos(child));
+        child->start_limit = child->stk.limit;
+        /*
+         * We push a fork point on the parent stack so we can teardown the child
+         * if we backtrack here.
+         */
+        extern struct jq_io_table jq__covt;
+        stack_push(jq, jq_handle_new(jq, "coroutine", &jq__covt, child));
+        stack_save(jq, pc - 1, stack_get_pos(jq));
+      } else {
+        /* Here we're in the child */
```

Comment:
I don't think this ever executes since the `initial_execution` hasn't been set to 1.
I believe if it was executed, we could get rid of one of the branches in the backtracking clause below

Reply:
alright, I get it now.
`jq_start` is being called with `reset`, where the `initial_execution` is properly set; in addition, it pushes a value on the stack.
But why isn't `jq_start` called after `coinit` for the initial run?